core/trino-main/src/main/java/io/trino/operator/aggregation/listagg/ListaggAggregationFunction.java

```diff
@@ -0,0 +1,335 @@
+        if (previousSeparator == null) {
+            state.setSeparator(otherState.getSeparator());
+            state.setOverflowError(otherState.isOverflowError());
+            state.setOverflowTruncationFiller(otherState.getOverflowTruncationFiller());
+            state.setOverflowTruncationCountIndication(otherState.isOverflowTruncationCountIndication());
+        }
+
+        state.merge(otherState);
+    }
+
+    public static void output(Type type, ListaggAggregationState state, BlockBuilder out)
+    {
+        if (state.isEmpty()) {
+            out.appendNull();
+        }
+        else {
+            outputState(state, out, MAX_OUTPUT_LENGTH);
+        }
+    }
+
+    @VisibleForTesting
+    protected static void outputState(ListaggAggregationState state, BlockBuilder out, int maxOutputLength)
+    {
+        Slice separator = state.getSeparator();
+        AtomicLong aggregationLength = new AtomicLong(0L);
+        AtomicInteger nonNullValuesCount = new AtomicInteger(0);
+        state.forEach((block, position) -> {
+            if (!block.isNull(position)) {
+                if (aggregationLength.get() > 0) {
+                    aggregationLength.addAndGet(separator.length());
+                }
+                aggregationLength.addAndGet(block.getSliceLength(position));
+                nonNullValuesCount.incrementAndGet();
+            }
+        });
+        if (nonNullValuesCount.get() == 0) {
+            out.appendNull();
+        }
+        else {
+            if (aggregationLength.get() > maxOutputLength) {
+                if (state.isOverflowError()) {
+                    throw new TrinoException(QUERY_REJECTED, "Concatenated string is too large");
+                }
+                else {
+                    outputTruncatedState(state, out, nonNullValuesCount.get(), maxOutputLength);
+                }
+            }
+            else {
+                AtomicBoolean elementWritten = new AtomicBoolean(false);
+                state.forEach((block, position) -> {
+                    if (!block.isNull(position)) {
+                        if (elementWritten.get()) {
+                            out.writeBytes(separator, 0, separator.length());
+                        }
+                        block.writeBytesTo(position, 0, block.getSliceLength(position), out);
+                        if (!elementWritten.get()) {
+                            elementWritten.set(true);
+                        }
+                    }
+                });
+            }
+            out.closeEntry();
+        }
+    }
+
+    private static void outputTruncatedState(ListaggAggregationState state, BlockBuilder out, int nonNullValuesCount, int maxOutputLength)
+    {
+        Slice separator = state.getSeparator();
+        Slice overflowTruncationFiller = state.getOverflowTruncationFiller();
+        AtomicInteger truncatedAggregationLength = new AtomicInteger(0);
+        AtomicInteger nonNullValuesWritten = new AtomicInteger(0);
```

Comment:
Similar comment as above regarding AtomicXXX

Reply:
As described here:
https://codippa.com/increment-variable-foreach-lambda-expression-java/
I am using a 1 item array to cope with this problem.

Reply:
I've broken the method `appendTruncatedBlock` into several other methods in an attempt to make the logic more readable.
Feedback is welcome.